util/hypercall_external_signal/orchestrator-request.py

```diff
@@ -0,0 +1,272 @@
+    """Receive a complete message from a socket connection.
+
+    Reads data in chunks until connection is closed or timeout occurs.
+    Handles messages larger than the standard buffer size.
+
+    Args:
+        conn: Connected socket to read from
+        timeout: Maximum time to wait for data in seconds (default: 30.0)
+
+    Returns:
+        str: Complete decoded message
+
+    Raises:
+        TimeoutError: If no data received within timeout period
+    """
+    chunks = []
+    while True:
+        ready, _, _ = select.select([conn], [], [], timeout)
+        if not ready:
+            raise TimeoutError("Timeout waiting for complete response")
+
+        chunk = conn.recv(4096)
+        if not chunk:  # Connection closed by remote
+            break
+        chunks.append(chunk)
+
+    return b"".join(chunks).decode()
+
+
+def send_and_receive_hypercall(pid: int, function: str) -> str:
+    """Send a hypercall to gem5 and wait for response.
+
+    Creates Unix socket, sends hypercall via transmitter, and waits for
+    response.
+
+    Args:
+        pid: Process ID of gem5 instance
+        function: Hypercall function to execute ('status'|'get_stats')
+
+    Returns:
+        str: JSON response from gem5
+
+    Raises:
+        TimeoutError: If gem5 doesn't respond within timeout
+        ValueError: If payload or response format is invalid
+    """
+    global sock, socket_path
+
+    socket_path = f"/tmp/hypercall_{os.getpid()}.sock"
+
+    try:
+        os.unlink(socket_path)
+    except OSError:
+        pass
+
+    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+    sock.bind(socket_path)
+    sock.listen(1)
+
+    try:
+        payload = json.dumps(
+            {"function": function, "response_socket": socket_path}
+        )
+        send_signal(pid, 1000, payload)
+
+        ready, _, _ = select.select([sock], [], [], 30.0)
+        if not ready:
+            raise TimeoutError("Timeout waiting for gem5 response")
+
+        conn, addr = sock.accept()
+        try:
+            return receive_full_message(conn)
+        finally:
+            conn.close()
+    finally:
+        cleanup()
+
+
+def write_response_to_file(response: str, filename: str) -> None:
+    """Write response to specified file.
+
+    Args:
+        response: JSON response string
+        filename: Path to output file
+    """
+    with open(filename, "w") as f:
+        f.write(response)
+
+
+def main():
+    signal.signal(signal.SIGINT, cleanup)
+
+    parser = argparse.ArgumentParser(description="Send hypercalls to gem5")
+    parser.add_argument(
+        "--pid",
+        type=int,
+        help="Process ID to send hypercall to "
+        "(auto-detected if not specified)",
+    )
+    parser.add_argument(
+        "function", choices=["status", "get_stats"], help="Function to execute"
+    )
+    parser.add_argument(
+        "--output", type=str, help="Write response to specified file"
```

Comment:
Is there much use for this? I would have assumed piping the output would work just as well.

Reply:
Yeah, I think so. For big files like stats it would be a bit of a pain. It doesn't hurt much to have it.